projects/ti_to_ti/pre_tokenize_parallel_example.py

```diff
@@ -0,0 +1,319 @@
+
+
+import io
+import os
+import json
+from abc import ABC, abstractmethod
+import copy
+from typing import Any, Callable
+from typing_extensions import TypedDict  # Python 3.10+
+from tqdm import tqdm
+
+import requests
+import librosa
+from PIL import Image
+from torchvision.io import read_video
+import uuid
+
+from align_anything.utils.template_registry import register_template
+
+
+import torch
+import torch.multiprocessing as mp
+
+import transformers
+from torch.utils.data import Dataset
+from torchvision import transforms
+from transformers.tokenization_utils import PaddingStrategy, TruncationStrategy
+from transformers import AutoTokenizer, ChameleonProcessor
+from align_anything.models.chameleon_model import AccustomedChameleonModel
+
+from align_anything.utils.multi_process import get_current_device
+from align_anything.utils.template_registry import get_template_class
+from align_anything.utils.tools import right_padding
+from datasets import load_dataset
+
+ALLOWED_ATTRIBUTES = ['split_token']
+DEFAULT_SPLIT_TOKEN = 'ASSISTANT:'
+IGNORE_INDEX = -100
+
+
+def load_image(image_path: str):
+    try:
+        if image_path.startswith("http"):
+            image = Image.open(requests.get(image_path, stream=True).raw)
+        else:
+            image = Image.open(image_path)
+        return image
+    except Exception as e:
+        print(f"Error occured when dealing with {image_path}")
+        raise Exception
+
+def format_sample_llava(raw_sample: dict[str, Any]) -> dict[str, Any]:
+    system_prompt: str = 'BEGINNING OF CONVERSATION: '
+    user_prompt: str = 'USER: \n<image>{input}'
+    assistant_prompt: str = '\nASSISTANT:{output}'
+    split_token: str = 'ASSISTANT:'
+    separator: str = '###'
+    raw_conversations = raw_sample['conversations']
+    raw_prompt = raw_conversations[0]['value'].replace('<image>\n', '').replace('\n<image>', '')
+
+    text = (
+        f'{system_prompt}'
+        f'{user_prompt.format(input=raw_prompt)}'
+        f"{assistant_prompt.format(output=raw_conversations[1]['value'])}"
+    )
+
+    prompt = (
+        f'{system_prompt}'
+        f'{user_prompt.format(input=raw_prompt)}'
+        f"{assistant_prompt.format(output='')}"
+    )
+
+    image_file = raw_sample['image_url']
+    return {
+        'text': text,
+        'prompt': prompt,
+        'input_image': [load_image(image_file)],
+        'image': [load_image(image_file)],
+    }
+
+
+def format_sample_cham(raw_sample: dict[str, Any]) -> dict[str, Any]:
+    system_prompt: str = 'BEGINNING OF CONVERSATION: '
+    user_prompt: str = 'USER: \n{input}'
+    assistant_prompt: str = '\nASSISTANT:{output}'
+    split_token: str = 'ASSISTANT:'
+    separator: str = '###'
+    # print(raw_sample)
+    input_text = raw_sample['question']
+    output_text = raw_sample['response']
+    input_img = raw_sample['image_url']
+    output_img = raw_sample['output_image_url']
+    
+    if isinstance(input_img, str):
+    # 单个图像
```

Comment:
remove this comment

Reply:
fixed in https://github.com/PKU-Alignment/align-anything/pull/41/commits/25f0b8d01ede5f8e0e5cb058a3607d356ae46724 commit